cmd/integration-test-server/main.go

```diff
@@ -0,0 +1,260 @@
+
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"math"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethstorage/go-ethstorage/cmd/es-utils/utils"
+	"github.com/ethstorage/go-ethstorage/ethstorage"
+	es "github.com/ethstorage/go-ethstorage/ethstorage"
+	"github.com/ethstorage/go-ethstorage/ethstorage/node"
+	prv "github.com/ethstorage/go-ethstorage/ethstorage/prover"
+)
+
+const (
+	expectedSaidHelloTime    = 10 * time.Minute
+	expectedStateRefreshTime = 5 * time.Minute
+	executionTime            = 2 * time.Hour
+
+	kvEntries = 32768
+	kvSize    = 128 * 1024
+	dataSize  = 126976
+
+	blobEmptyFillingMask = byte(0b10000000)
+
+	uploadedDataFile = ".data"
+	shardFile        = "../../es-data-it/shard-0.dat"
+)
+
+var (
+	portFlag = flag.Int("port", 9096, "Listener port for the es-node to report node status")
+)
+
+var (
+	errorMessages    = make([]string, 0)
+	lastQueryTime    = time.Now()
+	lastRecord       *node.NodeState
+	hasConnectedPeer = false
+	testLog          = log.New("IntegrationTest")
+	prover           = prv.NewKZGProver(testLog)
+)
+
+func HelloHandler(w http.ResponseWriter, r *http.Request) {
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		addErrorMessage(fmt.Sprintf("Read Hello body failed with error %s", err.Error()))
+		return
+	}
+	log.Info("Get hello from node", "id", string(body))
+
+	if time.Since(lastQueryTime) > expectedSaidHelloTime {
+		addErrorMessage(fmt.Sprintf("Get Hello message later then expect time %v real value %v", expectedSaidHelloTime, time.Since(lastQueryTime)))
+	}
+	lastQueryTime = time.Now()
+	log.Info("Get Hello request from es-node", "id", string(body))
+
+	answer := `{"status":"ok"}`
+	w.Write([]byte(answer))
+}
+
+func ReportStateHandler(w http.ResponseWriter, r *http.Request) {
+	if time.Since(lastQueryTime) > expectedStateRefreshTime*2 {
+		addErrorMessage(fmt.Sprintf("Get Hello message later then expect time %v real value %v",
+			expectedStateRefreshTime*2, time.Since(lastQueryTime)))
+	}
+	lastQueryTime = time.Now()
+
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		addErrorMessage(fmt.Sprintf("Read ReportState body failed with error %s", err.Error()))
+		return
+	}
+
+	state := &node.NodeState{}
+	err = json.Unmarshal(body, state)
+	if err != nil {
+		addErrorMessage(fmt.Sprintf("Parse node state failed with error %s", err.Error()))
+		w.Write([]byte(fmt.Sprintf(`{"status":"error", "err message":"%s"}`, err.Error())))
+		return
+	}
+
+	// If no state updated
+	log.Info("Get state from peer", "peer id", state.Id, "Version", state.Version)
+	if lastRecord != nil {
+		checkState(lastRecord, state)
+	}
+
+	lastRecord = state
+
+	w.Write([]byte(`{"status":"ok"}`))
+}
+
+func checkState(oldState, newState *node.NodeState) {
+	if len(oldState.Shards) != len(newState.Shards) {
```

Comment:
Should we consider the case of opening new shards?

Reply:
We cannot add or remove shards while running es-node, right? We can only update the node's shards by sharding it down and restarting it again.

Reply:
Actually we can init the shards beforehand in the local es-node, but it depends on the requirements of the tests.

Reply:
let's add after multi-shards support PR check-in?

Reply:
If we agree that this is the case we want to include in the integration test, I think you should add it to the checkpoints in the PR description. Also, please leave a note mentioning that we will need a separate PR to add this case if we have to wait for the multi-shard support PR to be merged

Reply:
add TODO for this.